components/application-broker/cmd/eventing-service-instance-migration/serviceinstance.go

```diff
@@ -0,0 +1,330 @@
+	istioauthenticationv1alpha1apis "istio.io/client-go/pkg/apis/authentication/v1alpha1"
+	istioclientset "istio.io/client-go/pkg/clientset/versioned"
+)
+
+const (
+	appBrokerServiceClass         = "application-broker"
+	eventsServiceEntryType        = "Events"
+	policyKnativeBrokerLabelKey   = "eventing.knative.dev/broker"
+	policyKnativeBrokerLabelValue = "default"
+)
+
+type serviceInstancesList []servicecatalogv1beta1.ServiceInstance
+
+type eventActivationsByServiceInstance map[string]string
+type eventActivationsByServiceInstanceAndNamespace map[string]eventActivationsByServiceInstance
+type istioPolicyByNamespace = map[string]*istioauthenticationv1alpha1apis.Policy
+
+// serviceInstanceManager performs operations on ServiceInstances.
+type serviceInstanceManager struct {
+	svcCatalogClient servicecatalogclientset.Interface
+	dynClient        dynamic.Interface
+	istioClient      istioclientset.Interface
+
+	serviceInstances      serviceInstancesList
+	eventActivationsIndex eventActivationsByServiceInstanceAndNamespace
+	istioBrokerPolicies   istioPolicyByNamespace
+}
+
+// newServiceInstanceManager creates and initializes a serviceInstanceManager.
+func newServiceInstanceManager(svcCatalogClient servicecatalogclientset.Interface, dynClient dynamic.Interface,
+	istioClient istioclientset.Interface,
+	namespaces []string) (*serviceInstanceManager, error) {
+
+	m := &serviceInstanceManager{
+		svcCatalogClient: svcCatalogClient,
+		dynClient:        dynClient,
+		istioClient:      istioClient,
+	}
+
+	if err := m.populateServiceInstances(namespaces); err != nil {
+		return nil, errors.Wrap(err, "populating ServiceInstances cache")
+	}
+
+	if err := m.populateIstioPolicies(namespaces); err != nil {
+		return nil, errors.Wrap(err, "populating Istio Policy index")
+	}
+
+	return m, nil
+}
+
+// populateServiceInstances populates the local list of ServiceInstances.
+// In order to be marked as candidate for re-creation, a ServiceInstance must relate to a ServiceClass which:
+//  * is used for events
+//  * is handled by the Kyma Application broker
+func (m *serviceInstanceManager) populateServiceInstances(namespaces []string) error {
+	eventsSvcClasses, err := m.buildServiceClassIndex()
+	if err != nil {
+		return errors.Wrap(err, "building index of ServiceClasses")
+	}
+
+	svcBrokerIndex, err := m.buildServiceBrokerIndex(namespaces)
+	if err != nil {
+		return errors.Wrap(err, "building index of ServiceBrokers")
+	}
+
+	for _, ns := range namespaces {
+		svcis, err := m.svcCatalogClient.ServicecatalogV1beta1().ServiceInstances(ns).List(metav1.ListOptions{})
+		switch {
+		case apierrors.IsNotFound(err):
+			return NewTypeNotFoundError(err.(*apierrors.StatusError).ErrStatus.Details.Kind)
+		case err != nil:
+			return errors.Wrapf(err, "listing ServiceInstances in namespace %s", ns)
+		}
+
+		for _, svci := range svcis.Items {
+			// Eventing ServiceInstance will only have ServiceClassRef hence rest are ignored
```

Comment:
why do you think that if there is a serviceClassRef it meanest is eventing service instance?